csrc/aes_cbc.cpp

```diff
@@ -136,6 +152,68 @@ class AesCbcCipher {
         return result;
     }
 
+    int extended_update(bool is_iso10126_padding,
+        bool is_enc,
+        jbyteArray j_last_block,
+        uint8_t const* input,
+        int input_len,
+        uint8_t* output,
+        int unprocessed_input)
+    {
+        if (!is_iso10126_padding || is_enc) {
+            return update(input, input_len, output, unprocessed_input);
+        }
+        // We are decrypting and padding is ISO10126. In this case, we must not decrypt the last block until do_final.
+        // The last_block is used to store this information.
+        JBinaryBlob last_block(jenv_, nullptr, j_last_block);
+        // The last byte of last_block stores the number of bytes in it from the previous call.
+        int last_block_len = last_block.get()[AES_CBC_BLOCK_SIZE_IN_BYTES];
+        // This is the total number of bytes that we have. Some must be stored in the last_block and the rest must be
+        // passed to the cipher.
+        int all = last_block_len + input_len;
+        int rem = all % AES_CBC_BLOCK_SIZE_IN_BYTES;
+        // We need to store no more than 16 bytes of all the input. c holds the number of bytes that must be copied into
+        // last_block.
+        int c = rem == 0 ? min(AES_CBC_BLOCK_SIZE_IN_BYTES, all) : rem;
+        // p is the total number of bytes that must be passed to the cipher.
+        int p = all - c;
+        // p1 is the number of bytes that must be pass to the cipher from last_block.
+        int p1 = min(p, last_block_len);
+        // p2 is the number of bytes that must be pass to the cipher from input.
+        int p2 = p - p1;
+        // c2 is the number of bytes that must be copied from input to last_block.
+        int c2 = min(c, input_len);
+        // c1 is the number of bytes that must be copied from last_block to itself.
+        int c1 = c - c2;
+
+        // Processing step: passing bytes to the cipher. This includes passing bytes from last_block and input:
+        // Passing bytes from last_block:
+        int up = unprocessed_input - last_block_len;
+        int result = update(last_block.get(), p1, output, up);
+        up = (p1 + up) - result;
+
+        // Passing bytes from input
+        result += update(input, p2, output + result, up);
```

Comment:
Not quite, this approach forces all subsequent update calls to do two memory fetches from different segments and breaks spatial locality. This is true even if the caller performs their incremental updates on perfectly block aligned memory chunks because we always have to load that last 16-bytes from the scratch buffer.

Reply:
In this case, it's because we don't know if the current passed cipher text is all of it or more is to come.